src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java

```diff
@@ -1003,226 +997,99 @@ private String getPermissionFilterQuery(DataverseRequest dataverseRequest, SolrQ
             }
-           
+
             // ----------------------------------------------------
-            // (4) User is logged in AND onlyDatatRelatedToMe == true
+            // User is logged in AND onlyDatatRelatedToMe == true
             // Yes, give back everything -> the settings will be in
-            //          the filterqueries given to search
+            // the filterqueries given to search
             // ----------------------------------------------------
             if (onlyDatatRelatedToMe == true) {
                 if (systemConfig.myDataDoesNotUsePermissionDocs()) {
                     logger.fine("old 4.2 behavior: MyData is not using Solr permission docs");
-                    return dangerZoneNoSolrJoin;
+                    return buildPermissionFilterQuery(avoidJoin, ALL_GROUPS);
                 } else {
                     // fall-through
                     logger.fine("new post-4.2 behavior: MyData is using Solr permission docs");
                 }
             }
-            
             // ----------------------------------------------------
-            // (5) Work with Authenticated User who is not a Superuser
+            // Work with Authenticated User who is not a Superuser
             // ----------------------------------------------------
-
-            groups = groupService.collectAncestors(groupService.groupsFor(dataverseRequest));
+            groupList.add(IndexServiceBean.getGroupPerUserPrefix() + au.getId());
         }
         
-        if (FeatureFlags.AVOID_EXPENSIVE_SOLR_JOIN.enabled()) {
-            /**
-             * Instead of doing a super expensive join, we will rely on the 
-             * new boolean field PublicObject:true for public objects. This field 
-             * is indexed on the content document itself, rather than a permission 
-             * document. An additional join will be added only for any extra, 
-             * more restricted groups that the user may be part of.
-             * **Note the experimental nature of this optimization**. 
-             */
-            StringBuilder sb = new StringBuilder();
-            StringBuilder sbgroups = new StringBuilder();
-            
-            // All users, guests and authenticated, should see all the 
-            // documents marked as publicObject_b:true, at least:
-            sb.append(SearchFields.PUBLIC_OBJECT + ":" + true);
+        // In addition to the user referenced directly, we will also
+        // add joins on all the non-public groups that may exist for the
+        // user:
 
-            // One or more groups *may* also be available for this user. Once again,
-            // do note that Guest users may be part of some groups, such as 
-            // IP groups. 
-            
-            int groupCounter = 0;
+        // Authenticated users and GuestUser may be part of one or more groups; such
+        // as IP Groups.
```

Comment:
[nitpick] Not asking to revert this comment change, but just want to mention that I don't think the 2nd sentence there was meaningless or redundant. I thought it was slightly counter-intuitive and/or easy to forget that guest users could be parts of some group. Hence emphasizing it.
Maybe that sentence just needed an "also" - "may _also_ be part of some group".